src/common.rs

```diff
@@ -139,26 +151,28 @@ impl<'a, T, F> System<'a> for Merge<F>
                        WriteStorage<'a, F>,
                        WriteStorage<'a, T>);
 
-    fn run(&mut self, (entities, mut errors, mut future, mut pers): Self::SystemData) {
-        let mut delete = &mut self.tmp;
+    fn run(&mut self, (entities, mut errors, mut futures, mut pers): Self::SystemData) {
 
-        for (e, future) in (&*entities, &mut future).join() {
-            match future.poll() {
-                Ok(Async::Ready(x)) => {
-                    pers.insert(e, x);
-                    delete.push(e);
+        for (e, ()) in (&*entities, &futures.check()).join() {
+            self.spawns.push(RefCell::new((e, spawn(futures.remove(e).unwrap()))));
+        }
+
+        self.spawns.retain(|spawn| {
```

Comment:
In order to get rid of `RefCell`, we can use

```rust
let len = spawns.len();
let mut del = 0;
{
    for i in 0..len {
        let remove = { /* Do stuff */ };

        if remove {
            del += 1;
        } else if del > 0 {
            spawns.swap(i - del, i);
        }
    }
}
if del > 0 {
    spawns.truncate(len - del);
}
```

You can also just make this a helper function. The fact that `retain` only allows immutable access was a design mistake.

Reply:
Ha!  That's basically exactly what I wrote for SmallVec's retain function.  I love code re-use.

Reply:
That is almost like my first variant :smile: